bot.go

```diff
@@ -57,15 +60,50 @@ func handleDownload(b *bot.Bot, m *bot.Message) {
 }
 
 func handleSearch(b *bot.Bot, m *bot.Message, api *rarbg.API) {
-	re, _ := regexp.Compile("tt[0-9]{7}") // e.g. tt0137523
-	imdbIDs := re.FindAllString(m.Text, -1)
+	imdbIDs, err := searchIMDbIDsFromMessage(m.Text)
+	if err != nil {
+		b.Send(m.Sender, replyNoIMDbIDS+err.Error())
+		return
+	}
 	if len(imdbIDs) == 0 {
 		b.Send(m.Sender, helpText)
 		return
 	}
+
 	for _, id := range imdbIDs {
 		result := new(bytes.Buffer)
 		searchIMDb(result, id, api)
 		b.Send(m.Sender, result.String(), &bot.SendOptions{ParseMode: bot.ModeMarkdown})
 	}
 }
+
+func searchIMDbIDsFromMessage(text string) ([]string, error) {
+	imdbIDs := []string{}
+	// Douban
+	movieLinks := findDoubanMovieURLs(text)
+	for _, url := range movieLinks {
+		movie := douban.NewMovie()
+		if err := movie.FetchFromURL(url); err != nil {
+			return nil, err
+		}
+		imdbIDs = append(imdbIDs, movie.IMDbID())
+	}
+	// IMDB
+	if len(imdbIDs) == 0 {
+		imdbIDs = findIMDbIDs(text)
+	}
+	return imdbIDs, nil
+}
+
+var (
+	reDoubanMovieURL = regexp.MustCompile(`http(s)?\:\/\/movie\.douban\.com\/subject\/[0-9]+`)
```

Comment:
Should we support other URL forms? For example:

- https://www.douban.com/doubanapp/dispatch?uri=/movie/27027913/&dt_dapp=1 (douban.app shared links)
- https://m.douban.com/movie/subject/27027913/ (redirected links of douban.app shared links opened in safari)
- etc. (other situations may produce other forms)

To summarize, if a string contains both `douban.com` and `{subject_id}`, it can be a valid link. Can we still done it in one pass?

Reply:
If we support these forms of links, we may have to restuct them to “desktop links”.

Reply:
Well I didn't even know these kinds of links exist.

How about we follow all the links given in the message, and see if the final jump(301 & 302) could be the normal form?

Reply:
I just noticed Douban have some redirect policies for different UAs. We could address these situations later on.

Reply:
uh....

Reply:
These URLs all have 7 or 8 numbers (not sure about 6 or 9 etc.) and two slashes around them, how about using `/\d+/` to identify the subject id and fetch it using `https://movie.douban.com/subject/{id}/`?

It only cost a HTTP request, we can try it and tell user how does it going.

Any better idea?